design-documents/graph-ql/mutation-error-design.md

```diff
@@ -0,0 +1,153 @@
+# Mutation Error Design
+
+Mutations are _the_ single way to trigger side-effects in a GraphQL API. Side-effects are where the large majority of our errors are going to happen, as a result of changing some underlying data.
+
+Today, the Magento 2 GraphQL schema does not do a great job of describing what could go wrong as a result of running a mutation. We should fix that!
```

Comment:
While I see the value of providing a list of possible error codes in the schema (particularly for green field application), it may be problematic to extract all of them from the implementation of existing PHP-resolvers which rely on existing models and services. Those models and services do not properly declare exceptions in their signatures. It means that it will be extremely difficult to list all error codes for the specific mutation, we can discuss if partial list makes sense.
For the same reason I believe we cannot introduce a shared interface for all mutation output types.

In scope of https://github.com/magento/architecture/blob/master/design-documents/graph-ql/coverage/customer-reorder.graphqls we agreed that custom error types make sense for validation errors (when multiple errors can happen during the same mutation)

Reply:
> While I see the value of providing a list of possible error codes in the schema (particularly for green field application), it may be problematic to extract all of them from the implementation of existing PHP-resolvers which rely on existing models and services

Sure! I'm not suggesting we surface every possible error state known. But for many operations, the common set of failures is both obvious and relevant to a shopper. If we have `addProductToCart`, we know the obvious causes are bad sku and out of stock.

>  we can discuss if partial list makes sense.

Yep, that's what I'm advocating for!

> For the same reason I believe we cannot introduce a shared interface for all mutation output types

Provided a [sample schema](https://github.com/magento/architecture/pull/393#discussion_r453972900) in another comment. Let me know your thoughts.